ql/src/java/org/apache/hadoop/hive/ql/optimizer/SortedDynPartitionTimeGranularityOptimizer.java

```diff
@@ -0,0 +1,353 @@
+        } else {
+          descs.add(new ExprNodeColumnDesc(col.getTypeInfo(), ReduceField.VALUE.toString()+"."+colName, null, false));
+        }
+      }
+      RowSchema selRS = new RowSchema(granularitySelOp.getSchema());
+      SelectDesc selConf = new SelectDesc(descs, colNames);
+      SelectOperator backtrackSelOp = (SelectOperator) OperatorFactory.getAndMakeChild(
+          selConf, selRS, rsOp);
+
+      // Link backtrack SelectOp to FileSinkOp
+      fsOp.getParentOperators().clear();
+      fsOp.getParentOperators().add(backtrackSelOp);
+      backtrackSelOp.getChildOperators().add(fsOp);
+
+      // Update file sink descriptor
+      fsOp.getConf().setDpSortState(FileSinkDesc.DPSortState.PARTITION_SORTED);
+      fsOp.getConf().setPartitionCols(rsOp.getConf().getPartitionCols());
+      ColumnInfo ci = new ColumnInfo(granularitySelOp.getSchema().getSignature().get(
+              granularitySelOp.getSchema().getSignature().size() - 1)); // granularity column
+      fsOp.getSchema().getSignature().add(ci);
+
+      LOG.info("Inserted " + granularitySelOp.getOperatorId() + ", " + rsOp.getOperatorId() + " and "
+          + backtrackSelOp.getOperatorId() + " as parent of " + fsOp.getOperatorId()
+          + " and child of " + fsParent.getOperatorId());
+
+      parseCtx.setReduceSinkAddedBySortedDynPartition(true);
+      return null;
+    }
+
+    private Operator<? extends OperatorDesc> getGranularitySelOp(
+            Operator<? extends OperatorDesc> fsParent) throws SemanticException {
+      ArrayList<ColumnInfo> parentCols = Lists.newArrayList(fsParent.getSchema().getSignature());
+      ArrayList<ExprNodeDesc> descs = Lists.newArrayList();
+      List<String> colNames = Lists.newArrayList();
+      int timestampPos = -1;
+      for (int i = 0; i < parentCols.size(); i++) {
+        ColumnInfo ci = parentCols.get(i);
+        ExprNodeColumnDesc columnDesc = new ExprNodeColumnDesc(ci);
+        descs.add(columnDesc);
+        colNames.add(columnDesc.getExprString());
+        if (columnDesc.getTypeInfo().getCategory() == ObjectInspector.Category.PRIMITIVE
+                && ((PrimitiveTypeInfo) columnDesc.getTypeInfo()).getPrimitiveCategory() == PrimitiveCategory.TIMESTAMP) {
+          if (timestampPos != -1) {
+            throw new SemanticException("Multiple columns with timestamp type on query result; "
+                    + "could not resolve which one is the timestamp column");
+          }
+          timestampPos = i;
+        }
+      }
+      if (timestampPos == -1) {
+        throw new SemanticException("No column with timestamp type on query result; "
+                + "one column should be of timestamp type");
+      }
+      RowSchema selRS = new RowSchema(fsParent.getSchema());
+      // Granularity (partition) column
+      String udfName;
+      Class<? extends UDF> udfClass;
+      switch (parseCtx.getConf().getVar(HiveConf.ConfVars.HIVE_DRUID_INDEXING_GRANULARITY)) {
```

Comment:
@b-slim That is what we need to check here but we are currently not doing. I think it is because when I implemented this part, we did not know yet that we were going to define granularity per table basis too.

In this part of the code, we are just taking the granularity property from Hive configuration. We need to retrieve the value for the property from the table property (as you have done in other parts of the code); if it does not exist, then take it from the configuration (as in this line). Then, we execute the switch statement and the rest remains the same.